File: ptiles-v7/src/spatial_index.rs
/// Spatial Index (per layer)
///
/// H3 resolution 7 cells sorted by H3 cell ID for binary search.
/// Shared binary format with v6 (19 bytes per entry).
///
/// Layout:
///   entry_count (4 bytes, uint32)
///   For each entry:
///     h3_cell (8 bytes, uint64) - H3 index as integer
///     block_offset (6 bytes) - Absolute byte offset to data block
///     block_length (3 bytes) - Compressed block size
///     poi_count (2 bytes, uint16) - Features in this cell

use byteorder::{LittleEndian, ReadBytesExt};

pub const ENTRY_SIZE: usize = 19; // 8 + 6 + 3 + 2

#[derive(Debug, Clone)]
pub struct IndexEntry {
    pub h3_cell: u64,
    pub block_offset: u64,
    pub block_length: u32,
    pub poi_count: u16,
}

#[derive(Debug, Clone)]
pub struct SpatialIndex {
    pub entry_count: u32,
    pub entries: Vec<IndexEntry>,
}

impl SpatialIndex {
    /// Read spatial index from a reader at its current position.
    pub fn read<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
        let entry_count = reader.read_u32::<LittleEndian>()?;
        let mut entries = Vec::with_capacity(entry_count as usize);

        for _ in 0..entry_count {
            let h3_cell = reader.read_u64::<LittleEndian>()?;

            // Read 6-byte block_offset
            let mut offset_bytes = [0u8; 8];
            reader.read_exact(&mut offset_bytes[..6])?;
            let block_offset = u64::from_le_bytes(offset_bytes);

            // Read 3-byte block_length
            let mut len_bytes = [0u8; 4];
            reader.read_exact(&mut len_bytes[..3])?;
            let block_length = u32::from_le_bytes(len_bytes);

            let poi_count = reader.read_u16::<LittleEndian>()?;

            entries.push(IndexEntry {
                h3_cell,
                block_offset,
                block_length,
                poi_count,
            });
        }

        Ok(SpatialIndex {
            entry_count,
            entries,
        })
    }

    /// Binary search for an H3 cell in the index.
    pub fn find_cell(&self, h3_cell: u64) -> Option<&IndexEntry> {
        self.entries
            .binary_search_by_key(&h3_cell, |e| e.h3_cell)
            .ok()
            .map(|idx| &self.entries[idx])
    }

    /// Find all cells that intersect a bounding box (brute-force scan).
    /// For production use, a proper H3 polygon fill should be used instead.
    pub fn find_in_bounds(
        &self,
    ) -> impl Iterator<Item = &IndexEntry> {
        self.entries.iter()
    }
}

File: packages/ptiles-client/src/spatial-index.ts
// src/spatial-index.ts — spatial index parsing

import { BinaryReader } from './binary-reader.js';

export interface IndexEntry {
  h3_cell: bigint;
  block_offset: number;
  block_length: number;
  feature_count: number;
}

/**
 * Parse the spatial index from a buffer.
 * Format:
 *   u32 entry_count
 *   for each entry:
 *     u64  h3_cell
 *     u48  block_offset
 *     u24  block_length
 *     u16  feature_count
 */
export function parseIndex(buffer: Uint8Array): IndexEntry[] {
  const reader = new BinaryReader(buffer);
  const count = reader.readU32();
  const entries: IndexEntry[] = [];

  for (let i = 0; i < count; i++) {
    const h3Low = reader.readU32();
    const h3High = reader.readU32();
    const h3_cell = (BigInt(h3High) << 32n) | BigInt(h3Low);
    const block_offset = reader.readU48();
    const block_length = reader.readU24();
    const feature_count = reader.readU16();

    entries.push({
      h3_cell,
      block_offset,
      block_length,
      feature_count,
    });
  }

  return entries;
}

/**
 * Detect whether block offsets are relative or absolute.
 * Rule: relative = (first_entry.block_offset < header.blocks_offset) or entries is empty
 */
export function detectRelativeOffsets(
  entries: IndexEntry[],
  blocksOffset: number
): boolean {
  if (entries.length === 0) return true;
  return entries[0].block_offset < blocksOffset;
}

/**
 * Look up an H3 cell in the sorted index via binary search.
 * Returns the entry or null.
 */
export function lookupCell(
  entries: IndexEntry[],
  cell: bigint
): IndexEntry | null {
  let lo = 0;
  let hi = entries.length - 1;

  while (lo <= hi) {
    const mid = (lo + hi) >>> 1;
    const midCell = entries[mid].h3_cell;
    if (midCell === cell) {
      return entries[mid];
    } else if (midCell < cell) {
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }

  return null;
}

/**
 * Resolve the absolute file position for a block entry.
 */
export function resolveBlockOffset(
  entry: IndexEntry,
  blocksOffset: number,
  relative: boolean
): number {
  if (relative) {
    return blocksOffset + entry.block_offset;
  }
  return entry.block_offset;
}
